### ISSUE TITLE
TIFF with YCbCr JPEG data and wrong TIFFTAG_PHOTOMETRIC value

### ISSUE THREAD

**Issue by Schmidor** [haraldk/TwelveMonkeys#404]
The TIFF contains an Old-Style-JPEG with YCbCr data.
While the JPEG stream correctly identifies the color components this is overridden by TIFFTAG_PHOTOMETRIC as RGB.
[foto_0001.zip](https://github.com/haraldk/TwelveMonkeys/files/1660006/foto_0001.zip)

Is it possible to include a workaround for such JPEG TIFFs, and override the TIFF metadata if the photometric interpretation and JPEG color components do not match?

**Comment by haraldk** [comment]
Hmmm... 

In a way it's tempting to add a workaround. But the `PhotometricInterpretation` tag is clearly wrong here, and the spec is pretty clear on the subject:

> Implementers should note that many popular JPEG codecs (compressor/ decompressors) provide automatic color conversion and down sampling, so that the application may supply full-size RGB data which is nonetheless converted to down sampled YCbCr. This is an implementation convenience which does not excuse the TIFF control layer from its responsibility to know what is really going on. The Photometric Interpretation and sub sampling fields written to the file must describe what is actually in the file. 

To make matters worse, the JPEG stream doesn't really specify the "color space" used, it's inferred based on convention. The TIFF control structure should really be the authoritative source for this information.

Then again, RGB-encoded JPEGs are very, very rare. And this JPEG stream contains a JFIF segment, which only allows YCbCr or Grayscale. Component IDs follow spec, 1, 2, 3, which is an indication of YCbCr. But it's not subsampled, which normally is a slight indication of RGB. However, why would we trust software that violates the TIFF spec to follow the JFIF spec? :-P

FWIW, the TIFF displays the same in Windows Photoviewer as it does using our reader. I'll test OS X Preview as well. 

-- 
Harald K

**Comment by Schmidor** [comment]
I know the Windows Viewer does display it the same as our reader. 
The depressing part is that JAI and IrfanView display it as YCbCr.

I was mistakenly thinking the SOF contains different ID's for RGB and YCbCr. So I understand now why it is not that easy.

**Comment by haraldk** [comment]
Interestingly, ~OS X~ MacOS Preview did not display the TIFF at all. Maybe it has no support for Old JPEG/6..

Anyway, thought about it on the subway... I'm pretty sure that JAI (and maybe IrfanView) just use whatever CS they get from the JPEG decoder. I recently rewrote the color space detection algorithm for the `JPEGImageReader`, to closely match that of libJPEG, so it should be what most people expect. 

Perhaps the best solution would be to just issue a warning in the cases where the JPEG color space and `PhotometricInterpretation` does not match, and go with the CS from the JPEG? There really isn't a right or wrong here (it's the file that's incorrect), but we should probably do the same as "most other software"... I wonder what libTIFF does?

-- 
Harald K

**Comment by haraldk** [comment]
Okay,

I've implemented a fix for this. It tests okay for your image, but I don't have any others... 

The new color space detection code for determining JPEG color space is similar to the one I implemented in the `JPEGImageReader`, which again is very close to libJPEG. If the JPEG color space is compatible with the `PhotometricInterpretation` in the TIFF, no conversion is done in the TIFF reader, and we rely on the JPEG reader to do the correct conversion. Except for CMYK, which always needs the extra massaging.

It's not 100% bullet proof when the JPEG and the TIFF don't agree, but I chose to go with the TIFF for `Compression` JPEG/7 and trust the JPEG color space for Old-JPEG/6, if it is YCbCr as this is by far the most common. If this approach is proven to be too naive, we could have a look at the subsampling parameters of the JPEG stream, and if its in use, we can still be pretty sure it's YCbCr. But I left that out for now... :-)


-- 
Harald K
